feat(keyboard): allow to override keyboard handlers

barmac opened the pull request:
This PR allows to add listeners which override default keyboard handlers.

* use KeyboardEvent.key value by default
* allow to register KeyboardEvent listeners for specified keys
* allow to use priority for KeyboardEvents
* default listeners will use key values instead of keyCode
* preserve previous way of registering listeners

There are three ways to register a listener now using keyboard.addListener:
1. Pass key or array of keys, then priority and the callback
2. Pass key or array of keys, then callback (default priority will be used)
3. Pass sole callback for global keyboard listener (lower priority than the latter)

Closes #226

nikku commented on lib/features/keyboard/DiagramKeyBindings.js:
```diff
@@ -0,0 +1,134 @@
+ */
+export default function DiagramKeyBindings(keyboard, editorActions) {
+
+  var moveCanvasLeft = MoveCanvasFactory('left', keyboard._config, editorActions);
```
Why don't we plug in move canvas keyboard shortcuts via the MoveCanvas feature?

This would potentially allow us to easier override it (i.e. to restore the old behavior).

barmac replied:
It is actually done using the `MoveCanvas` feature, see: https://github.com/bpmn-io/diagram-js/pull/289/files#diff-ab467edcc671fb2f7e5cb7a91b06437d
The factory function was added to keep code DRY as the only difference between the functions responsible for the movement is the `direction` property passed to `MoveCanvas` feature. Alternatively, we could move the factory function to `DiagramKeyBindings.js` as a helper at the end of the file.

However, this does not seem to make it more difficult to override the listeners, because the way to do that is always the same: add listeners for the same keys but with higher priority.

Please let me know what you think.

philippfromme replied:
I'd also go for a `MoveCanvas` feature instead of a factory function.

nikku commented on lib/features/keyboard/Keyboard.js:
```diff
@@ -1,14 +1,40 @@
+  isArray,
+  isFunction,
+  map
+} from 'min-dash';
```
Separate import blocks via newline.

nikku commented on lib/features/keyboard/Keyboard.js:
```diff
@@ -22,50 +48,39 @@ import {
-  this._config = config || {};
+  this._config = assign({
+    speed: DEFAULT_SPEED,
+    invertY: false
```
As mentioned above, the move canvas specifics belong to a move canvas keybinding if you ask me.

As an alternative we could make the detection implicit, i.e. apply it once we detect MacOS.

barmac replied:
The `speed` was already there, but it wasn't clearly expressed in the config declaraction. As the fallback, the implementation used a magic number. Otherwise, we could have a separate configuration for each key binding. I think it is a good idea.

barmac replied:
I think a better place to move the config is in [#376](https://github.com/bpmn-io/bpmn-js/issues/376). I'll change it, but in another PR.

philippfromme replied:
Same here. Speed shouldn't be part of the generic keyboard configuration. Other features might implement something that they want to do at a certain speed. If the configured speed doesn't work for two independent features they can't make use of it anyway.

nikku commented on lib/features/keyboard/Keyboard.js:
```diff
@@ -91,6 +104,43 @@ Keyboard.$inject = [
+
+  // run global listeners only if no listeners for specific keys reacted
+  if (!eventBusResult) {
+    forEach(this._globalListeners, function(listener) {
```
Global listeners are being called with the code while local listeners are being called with the key? I'd rethink this behavior as I don't think it is intuitive.

barmac replied:
I agree that it's not intuitive. We use `keyCode` for `globalListeners` only in order to preserve the backward compatibility, i.e. to not break for example the [BpmnKeyBindings](https://github.com/bpmn-io/bpmn-js/blob/master/lib/features/keyboard/BpmnKeyBindings.js).
Alternatively, we could drop the old keyboard related behaviour as `KeyboardEvent.keyCode` is already a `deprecated` feature.

barmac replied:
There is no blog post regarding the way to add a keyboard listener ([one mention of the feature](https://bpmn.io/blog/posts/2016-copy-paste-search-bpmn-js-0-15.html)); the examples don't address this issue either. 
There are two forum topics regarding the keyboard issues: [1](https://forum.bpmn.io/t/copy-paste-and-keyboard-events/741/3), [2](https://forum.bpmn.io/t/copy-and-paste-keyboard-binding-and-eventbus-on-keyboard-event-register/2005)

The change to `key` only would break `BpmnKeyBindings` as it uses `keyCode`. However, the feature itself needs refactor as it is a one big listener for each of Bpmn-js-specific key bindings.

nikku commented on lib/features/keyboard/Keyboard.js:
```diff
@@ -1,14 +1,40 @@
  * and triggering diagram actions.
  *
+ * For pressed keys, keyboard fires an event according to following scheme:
+ * `keyboard.keydown.[event.key || event.keyCode]`
```
When will this be called with the keyCode? I.e. how to separate the `1` key from the `1` keyCode?

barmac replied:
Removed the fallback.

nikku commented on lib/features/keyboard/DiagramKeyBindings.js:
```diff
@@ -0,0 +1,134 @@
+  var moveCanvasRight = MoveCanvasFactory('right', keyboard._config, editorActions);
+  var moveCanvasDown = MoveCanvasFactory('down', keyboard._config, editorActions);
+
+  keyboard.addListener(copyKeys, DEFAULT_PRIORITY, copy);
```
Could we support something event more powerful for our users here:

```
keyboard.addCombo([ `ctrl+A`, `ArrowUp` ], doSomething)
```

This would be a true simplification for the API.

Inspired by [hotshot](https://github.com/Shopify/hotshot).

barmac replied:
Thank you, I will have a look at that.

barmac commented:
Thank you for the feedback @nikku. 

Indeed, the PR seems to add a lot of code to the library. However, the previous implementation seemed to rely too much on hardcoded order of listeners, that's why I decided to use `EventBus` instead. Probably I should just change the PR's name. What do you think?

barmac commented:
Removed `keyCode` fallback for the new-style listeners. Linted the block import statement.

nikku commented:
If you want to move fast I suggest you to only apply the relevant changes to get priorities to the old behavior.

It should be trivial to add priorities to the old [Keyboard#addListener](https://github.com/bpmn-io/diagram-js/blob/master/lib/features/keyboard/Keyboard.js#L299) implementation. Even implementing the whole listener logic via the eventBus should be an easy and straight forward thing to do.

Next up we could take our time to think about whether a better keyboard API makes sense, introduce breaking changes and so forth.

nikku commented:
Imagine every key event went via the eventBus. In order to get the hotshot behavior a simple addition could be a utility that matches key strokes (or event sequences) using a pattern:

```javascript
// add a cut listener with high priority
keyboard.addListener(5000, combo('ctrl+x', function(event) {
  // `combo` utility ensures that this method is only called if key stroke matches combo
}));

// add listener for sequence of events
keyboard.addListener(5000, sequence('a b', function(event) {
  // `sequence` utility ensures that this method is only called if two keys in sequence matched <a> and <b>
}));

// or register via eventBus directly

// do we actually need a distinction between key{up|down}?
// or is it enough to support a simple key event?
eventBus.on('keyboard.key{up,down}', 5000, function(event) {
  // called on any key; may be wrapped to filter only relevant invocations
});
```

nikku commented:
I totally agree separating the keyboard from the default key events. This should be a separate commit though, as it is a clear non-breaking refactoring.

barmac commented:
> Imagine every key event went via the eventBus. In order to get the hotshot behavior a simple addition could be a utility that matches key strokes (or event sequences) using a pattern:
> 
> ```js
> // add a cut listener with high priority
> keyboard.addListener(5000, combo('ctrl+x', function(event) {
>   // `combo` utility ensures that this method is only called if key stroke matches combo
> }));
> 
> // add listener for sequence of events
> keyboard.addListener(5000, sequence('a b', function(event) {
>   // `sequence` utility ensures that this method is only called if two keys in sequence matched <a> and <b>
> }));
> 
> // or register via eventBus directly
> 
> // do we actually need a distinction between key{up|down}?
> // or is it enough to support a simple key event?
> eventBus.on('keyboard.key{up,down}', 5000, function(event) {
>   // called on any key; may be wrapped to filter only relevant invocations
> });
> ```

Okay, but this way we always trigger all (or almost all) the event listeners and let them pass silently, if they are not interested in the pressed key, correct?

nikku commented:
> Okay, but this way we always trigger all (or almost all) the event listeners and let them pass silently, if they are not interested in the pressed key, correct?

Yes. Did you benchmark if this is an issue? I assume it is not.

barmac commented:
I've rewritten this PR so that we use `EventBus` for all events, introducing this way a breaking change to `Keyboard`. The listeners will be able to decide whether to react basing on `keyCode` or `key` or both as they will receive `KeyboardEvent`. Also, the change is now split up between two commits: one for the breaking change, another for moving the default listeners to a separate file.

barmac commented on lib/features/keyboard/Keyboard.js:
```diff
@@ -34,38 +51,19 @@ export default function Keyboard(config, eventBus, editorActions) {
     self.unbind();
-    self._listeners = null;
+
+    forEach(self._listeners, function(listener) {
```
OK, then I'll remove that.

barmac commented:
Applied changes suggested by @nikku. Thanks for the feedback :)

barmac commented:
Added missing dependency.
